### PR TITLE
LDAP: Add API endpoint to debug user mapping from LDAP

### PR DESCRIPTION
**Pull request by gotjosh**
This endpoint returns a user found within the configured LDAP server(s). Moreso, it provides the mapping information for the user to help administrators understand how the users would be created within Grafana based on the current configuration.

No changes are executed or saved to the database, this is all an in-memory representation of how the final result would look like.

Best viewed commit-by-commit.

### CODE REVIEW

File: pkg/api/ldap_debug.go
Comment (gotjosh) [inline pkg/api/ldap_debug.go]: This is only relevant for Enterprise and likely to change _very_ soon once I put the other PR for that. Please avoid putting too much emphasis on that right now.

File: pkg/services/multildap/testing.go
Comment (gotjosh) [inline pkg/services/multildap/testing.go]: I ended up doing [shotgun surgery](https://refactoring.guru/smells/shotgun-surgery) by having to change three different mocks in three different places. Generally, it is a strong indication that something is wrong with the current state of these LDAP related tests but I decided to leave that for later (if that) - I will probably come back to this sometime this week and if I choose not to touch it we can decide whenever we can live with it or should we create an issue for it.

File: pkg/api/ldap_debug.go
Comment (torkelo) [inline pkg/api/ldap_debug.go]: team name?

**Review by torkelo — APPROVED** [review]
Looks good!

File: pkg/api/ldap_debug.go
Comment (gotjosh) [reply pkg/api/ldap_debug.go]: Interesting. I took inspiration from our current users API  (`/api/users`) where we only expose the `OrgId` - I assumed this is for issuing a separate request on the front-end for each organisation.

Is that not the case? I'm happy to change the structure and make the queries to the db right here but not sure that's what we do in other areas.

File: pkg/services/multildap/multildap.go
Comment (xlson) [inline pkg/services/multildap/multildap.go]: I'm thinking that it might be a good idea not to return a pointer here. In the current codebase we never re-use the `ldap.ServerConfig` as far as I'm aware, so there shouldn't be any problems. But I could easily see us changing that at some point and then running into problems with a return ServerConfig struct that is modified. I think it's fine to not be defensive while send the config into the LDAP module, you kinda expect whoever is setting up LDAP to not modify the config after its been entered. But I wouldn't neccessarily make the assumptions with config that is returned from LDAP.

File: pkg/api/ldap_debug.go
Comment (xlson) [inline pkg/api/ldap_debug.go]:
```suggestion
		return Error(http.StatusBadRequest, "Validation error. You must specify a username", nil)
```

I think a is correct here

**Review by xlson — APPROVED** [review]
LGTM!

File: pkg/api/ldap_debug.go
Comment (torkelo) [reply pkg/api/ldap_debug.go]: Not sure I understand you. Without the team name how can we show this in the table in the UI?

File: pkg/services/multildap/multildap.go
Comment (gotjosh) [reply pkg/services/multildap/multildap.go]: I agree. The reason why I left it as is due to the place of usage. 

At the moment of writing, the only place where we use the config is for displaying information as part of a short-lived HTTP request and I felt it doesn't really make a difference in there. The chances of a race are very slim IMO.

**Comment by xlson** [comment]
@gotjosh don't forget to set the milestone and add your PR to the project board

File: pkg/services/multildap/multildap.go
Comment (xlson) [reply pkg/services/multildap/multildap.go]: I agree that the risk for problems right now is petty much zero. The problem we could get into is if we ever change it so that we don't re-read the config every time we set up the LDAP services. If we ever start re-using the config we could run into non-obvious problems with this code.

File: pkg/api/ldap_debug.go
Comment (gotjosh) [reply pkg/api/ldap_debug.go]: We had a clear-up on Slack about this one. The summary of the discussion is: 

> Generally, for list APIs we usually try to include everything needed by UI

File: pkg/services/multildap/multildap.go
Comment (gotjosh) [reply pkg/services/multildap/multildap.go]: Good shout. I have changed the return to a value.

**Comment by gotjosh** [comment]
@torkelo and @xlson I know you both have approved the PR but wouldn't hurt a second pair of eyes on the last 6 commits (they are relatively small) before I merge this. Is there any chance I can get a quick re-review?

**Comment by torkelo** [comment]
Looks good!

**Review by xlson — APPROVED** [review]
Looks good!

**Pull request merged**
